runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java

```diff
@@ -243,6 +249,7 @@ private synchronized void setState(JobState.Enum state) {
   }
 
   private synchronized void sendMessage(JobMessage message) {
+    messageHistory.add(message);
```

Comment:
I've added this to make it consistent with the state updates, which we also save. However, this may be too memory intense, I'm seeing that the test time increased here: https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Commit/2320/

Need to check if this is due to this change. If so, we could subscribe to the messages before running the pipeline which would remove the need to save the messages.

Reply:
yeah, I'd be wary of storing all of that.  There should be a way to cap the number of historical messages a user cares about.   We could choose a sensible default for now and come back and expose it as an option later if anyone cares to do so.

Reply:
Right now, it looks like sendMessage is only being called by onFailure:

https://github.com/apache/beam/blob/1db2e3976fb2a5c15c77e55928e19c003db3b387/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java#L141-L150

Meaning there should be a maximum of 1 message. Unless there's something I'm missing?

Reply:
Good point.  I can confirm that is the behavior that I have seen.

Reply:
> yeah, I'd be wary of storing all of that.

We have the same problem for the state updates which are smaller but also can arrive at a higher frequency. How about (1) limiting the number of past invocations? We would keep all the state updates / messages per invocation. (2) An alternative could be to just store the error-related messages. (3) A third option would be to subscribe to the message stream early before pipeline execution starts.

I'm leaning towards (1) because that would still allow the retrieval of past state updates that I think was a requirement for you @chadrik.

>Meaning there should be only 2 messages per pipeline failure. Unless there's something I'm missing?

That's right. However, the messages also contain the state updates which we already save. So there is some room for optimization the storage.

Will follow-up regarding the build time but it looks like it almost doubled which we should definitely take serious.